# Backport of JDK-8275509 and its follow-up JDK-8290041:
#   8275509: ModuleDescriptor.hashCode isn't reproducible across builds
#   8290041: ModuleDescriptor.hashCode is inconsistent
#
# ModuleDescriptor.hashCode() and the hashCode() of Requires, Exports and
# Opens folded in the hash of their Set of modifiers. Those are sets of
# enums, and Enum.hashCode() is the identity hash code, which depends on
# the state of HotSpot's per-thread PRNG when the constant is first hashed
# - so it varies between JVM runs, and with it the descriptor hash.
#
# jlink bakes ModuleDescriptor.hashCode() into the SystemModules classes it
# generates (SystemModulesPlugin.putModuleDescriptor calls
# mv.visitLdcInsn(md.hashCode())), so this landed in lib/modules as an int
# constant that changed from build to build. It was the last remaining
# difference between two openSUSE builds, and it is sensitive to anything
# that shifts execution order in the jlink JVM - such as the number of
# build cores.
#
# Hash the modifier names instead of the enum instances. JDK-8290041 is the
# follow-up that makes the sum order independent again, which matters here
# because the modifier sets are Set.of instances whose iteration order is
# also randomized per JVM run - without it the first fix alone would leave
# the same symptom.
#
# Note that jlink runs from the JDK being built (BUILD_JDK defaults to
# JDK_OUTPUTDIR for a native build), so patching java.base takes effect for
# the image built in the same run.
#

diff --git a/src/java.base/share/classes/java/lang/module/ModuleDescriptor.java b/src/java.base/share/classes/java/lang/module/ModuleDescriptor.java
index fa359348f6c..f9ebfe1559d 100644
--- a/src/java.base/share/classes/java/lang/module/ModuleDescriptor.java
+++ b/src/java.base/share/classes/java/lang/module/ModuleDescriptor.java
@@ -333,7 +333,7 @@ public class ModuleDescriptor
          */
         @Override
         public int hashCode() {
-            int hash = name.hashCode() * 43 + mods.hashCode();
+            int hash = name.hashCode() * 43 + modsHashCode(mods);
             if (compiledVersion != null)
                 hash = hash * 43 + compiledVersion.hashCode();
             if (rawCompiledVersion != null)
@@ -514,7 +514,7 @@ public class ModuleDescriptor
          */
         @Override
         public int hashCode() {
-            int hash = mods.hashCode();
+            int hash = modsHashCode(mods);
             hash = hash * 43 + source.hashCode();
             return hash * 43 + targets.hashCode();
         }
@@ -721,7 +721,7 @@ public class ModuleDescriptor
          */
         @Override
         public int hashCode() {
-            int hash = mods.hashCode();
+            int hash = modsHashCode(mods);
             hash = hash * 43 + source.hashCode();
             return hash * 43 + targets.hashCode();
         }
@@ -2283,7 +2283,7 @@ public class ModuleDescriptor
         int hc = hash;
         if (hc == 0) {
             hc = name.hashCode();
-            hc = hc * 43 + Objects.hashCode(modifiers);
+            hc = hc * 43 + modsHashCode(modifiers);
             hc = hc * 43 + requires.hashCode();
             hc = hc * 43 + Objects.hashCode(packages);
             hc = hc * 43 + exports.hashCode();
@@ -2568,6 +2568,18 @@ public class ModuleDescriptor
                 .collect(Collectors.joining(" "));
     }
 
+    /**
+     * Generates and returns a hashcode for the enum instances. The returned hashcode
+     * is a value based on the {@link Enum#name() name} of each enum instance.
+     */
+    private static int modsHashCode(Iterable<? extends Enum<?>> enums) {
+        int h = 0;
+        for (Enum<?> e : enums) {
+            h += e.name().hashCode();
+        }
+        return h;
+    }
+
     private static <T extends Object & Comparable<? super T>>
     int compare(T obj1, T obj2) {
         if (obj1 != null) {
diff --git a/test/jdk/java/lang/module/ModuleDescriptorHashCodeTest.java b/test/jdk/java/lang/module/ModuleDescriptorHashCodeTest.java
new file mode 100644
index 00000000000..c1269f28c48
--- /dev/null
+++ b/test/jdk/java/lang/module/ModuleDescriptorHashCodeTest.java
@@ -0,0 +1,190 @@
+/*
+ * Copyright (c) 2021, 2022, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.module.ModuleDescriptor;
+import java.lang.module.ModuleDescriptor.Exports;
+import java.lang.module.ModuleDescriptor.Opens;
+import java.lang.module.ModuleDescriptor.Requires;
+import java.util.Set;
+
+import org.testng.annotations.Test;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotSame;
+
+/**
+ * @test
+ * @bug 8275509 8290041
+ * @summary Tests the ModuleDescriptor.hashCode()
+ * @run testng ModuleDescriptorHashCodeTest
+ * @run testng/othervm -Xshare:off ModuleDescriptorHashCodeTest
+ */
+public class ModuleDescriptorHashCodeTest {
+
+    /**
+     * Verifies that the ModuleDescriptor.hashCode() returned by a boot layer module is
+     * the same as that returned by a ModuleDescriptor constructed from the ModuleDescriptor.Builder
+     * for the same module.
+     */
+    @Test
+    public void testBootModuleDescriptor() throws Exception {
+        Set<Module> bootModules = ModuleLayer.boot().modules();
+        for (Module bootModule : bootModules) {
+            System.out.println("Testing module descriptor of boot module " + bootModule);
+            ModuleDescriptor bootMD = bootModule.getDescriptor();
+            ModuleDescriptor mdFromBuilder = fromModuleInfoClass(bootModule);
+            // verify that this object is indeed a different object instance than the boot module descriptor
+            // to prevent any artificial passing of the test
+            assertNotSame(mdFromBuilder, bootMD, "ModuleDescriptor loaded from boot layer and " +
+                    "one created from module-info.class unexpectedly returned the same instance");
+            assertEquals(mdFromBuilder.hashCode(), bootMD.hashCode(),
+                    "Unexpected ModuleDescriptor.hashCode() for " + mdFromBuilder);
+            assertEquals(mdFromBuilder.compareTo(bootMD), 0,
+                    "Unexpected ModuleDescriptor.compareTo() for " + mdFromBuilder);
+        }
+    }
+
+    /**
+     * Verifies that two "equal" module descriptors which only differ in the order of
+     * {@link ModuleDescriptor.Opens.Modifier opens modifiers}, that were used to construct the
+     * descriptors, have the same hashcode.
+     */
+    @Test
+    public void testOpensModifiersOrdering() throws Exception {
+        // important to use Set.of() (i.e. backed by immutable set) to reproduce the issue
+        final Set<Opens.Modifier> mods1 = Set.of(Opens.Modifier.SYNTHETIC, Opens.Modifier.MANDATED);
+        final ModuleDescriptor desc1 = createModuleDescriptor(mods1, null, null);
+
+        // create the same module descriptor again and this time just change the order of the
+        // "opens" modifiers' Set.
+
+        // important to use Set.of() (i.e. backed by immutable set) to reproduce the issue
+        final Set<Opens.Modifier> mods2 = Set.of(Opens.Modifier.MANDATED, Opens.Modifier.SYNTHETIC);
+        final ModuleDescriptor desc2 = createModuleDescriptor(mods2, null, null);
+
+        // basic verification of the modifiers themselves before we check the module descriptors
+        assertEquals(mods1, mods2, "Modifiers were expected to be equal");
+
+        // now verify the module descriptors
+        assertEquals(desc1, desc2, "Module descriptors were expected to be equal");
+        assertEquals(desc1.compareTo(desc2), 0, "compareTo was expected to return" +
+                " 0 for module descriptors that are equal");
+        System.out.println(desc1 + " hashcode = " + desc1.hashCode());
+        System.out.println(desc2 + " hashcode = " + desc2.hashCode());
+        assertEquals(desc1.hashCode(), desc2.hashCode(), "Module descriptor hashcodes" +
+                " were expected to be equal");
+    }
+
+    /**
+     * Verifies that two "equal" module descriptors which only differ in the order of
+     * {@link ModuleDescriptor.Exports.Modifier exports modifiers}, that were used to construct the
+     * descriptors, have the same hashcode.
+     */
+    @Test
+    public void testExportsModifiersOrdering() throws Exception {
+        // important to use Set.of() (i.e. backed by immutable set) to reproduce the issue
+        final Set<Exports.Modifier> mods1 = Set.of(Exports.Modifier.SYNTHETIC, Exports.Modifier.MANDATED);
+        final ModuleDescriptor desc1 = createModuleDescriptor(null, null, mods1);
+
+        // create the same module descriptor again and this time just change the order of the
+        // "exports" modifiers' Set.
+
+        // important to use Set.of() (i.e. backed by immutable set) to reproduce the issue
+        final Set<Exports.Modifier> mods2 = Set.of(Exports.Modifier.MANDATED, Exports.Modifier.SYNTHETIC);
+        final ModuleDescriptor desc2 = createModuleDescriptor(null, null, mods2);
+
+        // basic verification of the modifiers themselves before we check the module descriptors
+        assertEquals(mods1, mods2, "Modifiers were expected to be equal");
+
+        // now verify the module descriptors
+        assertEquals(desc1, desc2, "Module descriptors were expected to be equal");
+        assertEquals(desc1.compareTo(desc2), 0, "compareTo was expected to return" +
+                " 0 for module descriptors that are equal");
+        System.out.println(desc1 + " hashcode = " + desc1.hashCode());
+        System.out.println(desc2 + " hashcode = " + desc2.hashCode());
+        assertEquals(desc1.hashCode(), desc2.hashCode(), "Module descriptor hashcodes" +
+                " were expected to be equal");
+    }
+
+    /**
+     * Verifies that two "equal" module descriptors which only differ in the order of
+     * {@link ModuleDescriptor.Requires.Modifier requires modifiers}, that were used to construct the
+     * descriptors, have the same hashcode.
+     */
+    @Test
+    public void testRequiresModifiersOrdering() throws Exception {
+        // important to use Set.of() (i.e. backed by immutable set) to reproduce the issue
+        final Set<Requires.Modifier> mods1 = Set.of(Requires.Modifier.SYNTHETIC, Requires.Modifier.MANDATED);
+        final ModuleDescriptor desc1 = createModuleDescriptor(null, mods1, null);
+
+        // create the same module descriptor again and this time just change the order of the
+        // "exports" modifiers' Set.
+
+        // important to use Set.of() (i.e. backed by immutable set) to reproduce the issue
+        final Set<Requires.Modifier> mods2 = Set.of(Requires.Modifier.MANDATED, Requires.Modifier.SYNTHETIC);
+        final ModuleDescriptor desc2 = createModuleDescriptor(null, mods2, null);
+
+        // basic verification of the modifiers themselves before we check the module descriptors
+        assertEquals(mods1, mods2, "Modifiers were expected to be equal");
+
+        // now verify the module descriptors
+        assertEquals(desc1, desc2, "Module descriptors were expected to be equal");
+        assertEquals(desc1.compareTo(desc2), 0, "compareTo was expected to return" +
+                " 0 for module descriptors that are equal");
+        System.out.println(desc1 + " hashcode = " + desc1.hashCode());
+        System.out.println(desc2 + " hashcode = " + desc2.hashCode());
+        assertEquals(desc1.hashCode(), desc2.hashCode(), "Module descriptor hashcodes" +
+                " were expected to be equal");
+    }
+
+    // Returns a ModuleDescriptor parsed out of the module-info.class of the passed Module
+    private static ModuleDescriptor fromModuleInfoClass(Module module) throws IOException {
+        try (InputStream moduleInfo = module.getResourceAsStream("module-info.class")) {
+            if (moduleInfo == null) {
+                throw new RuntimeException("Could not locate module-info.class in " + module);
+            }
+            // internally calls ModuleDescriptor.Builder
+            return ModuleDescriptor.read(moduleInfo);
+        }
+    }
+
+    // creates a module descriptor with passed (optional) opens/exports/requires modifiers
+    private static ModuleDescriptor createModuleDescriptor(
+            Set<Opens.Modifier> opensModifiers,
+            Set<Requires.Modifier> reqsModifiers,
+            Set<Exports.Modifier> expsModifiers) {
+
+        final ModuleDescriptor.Builder builder = ModuleDescriptor.newModule("foobar");
+        if (opensModifiers != null) {
+            builder.opens(opensModifiers, "a.p1", Set.of("a.m1"));
+        }
+        if (reqsModifiers != null) {
+            builder.requires(reqsModifiers, "a.m2");
+        }
+        if (expsModifiers != null) {
+            builder.exports(expsModifiers, "a.b.c", Set.of("a.m3"));
+        }
+        return builder.build();
+    }
+}
